### PR TITLE
Fixed bug where new folders could destroy existing folders with same name

### PR DESCRIPTION
**Pull request by bsclifton**
- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).

Test Plan:

Fixes https://github.com/brave/browser-laptop/issues/4728

Auditors: @darkdh

Test Plan:
1. Create a bookmark folder named 'x'
2. Add a bookmark to the folder
3. Create new bookmark folder named 'x'
4. Check the contents of the folder

### CODE REVIEW

File: js/state/siteUtil.js [hunk 1]
```diff
@@ -101,65 +145,36 @@ module.exports.addSite = function (sites, siteDetail, tag, originalSiteDetail) {
...
-  let site = Immutable.fromJS({
-    lastAccessedTime: siteDetail.get('lastAccessedTime') || new Date().getTime(),
-    tags,
-    title: siteDetail.get('title')
-  })
-  if (siteDetail.get('location')) {
-    site = site.set('location', siteDetail.get('location'))
-  }
+  let site = mergeSiteDetails(oldSite, siteDetail, tag)
   if (folderId) {
```
Comment (darkdh) [inline js/state/siteUtil.js]: Can we move this into `mergeSiteDetails `?

Comment (bsclifton) [inline js/state/siteUtil.js]: Done :smile:

**Comment by alexwykoff** [comment]
Do we feel confident that this will cover the import case as well?

File: test/unit/state/siteUtilTest.js [hunk 2]
```diff
@@ -154,24 +154,122 @@ describe('siteUtil', function () {
...
+            parentFolderId: folderId,
+            location: testUrl1,
+            tags: [siteTags.BOOKMARK]
+          })
+          // Add a bookmark into that folder
+          processedSites = siteUtil.addSite(processedSites, bookmark)
+          assert.equal(processedSites.size, 2)
+          assert.equal(processedSites.getIn([1, 'parentFolderId']), folderId)
+
+          // Add another bookmark folder with the same name / parentFolderId
```
Comment (darkdh) [inline test/unit/state/siteUtilTest.js]: great coverage!

**Comment by darkdh** [comment]
@alexwykoff it can cover import same data repeatedly

**Comment by darkdh** [comment]
lgtm!

**Comment by darkdh** [comment]
Test plan updated:
1. Import bookmarks from other browsers (must have folder 'x' on bookmark toolbar and contents in it)
2. Create a new folder with the name 'x'  on bookmark toolbar
3. The imported folder shouldn't be replaced
4. Add a bookmark in folder 'x' you created
5. Do the importing process as step 1 again
6. The folder created in step 2 and bookmark in step 4 shouldn't be deleted

**Comment by ivandejanovic** [comment]
This issue seems to exists when importing bookmarks that contain to folders with same name on different paths.

Replication steps:

Create this bookmark structure in some browser. (I used Firefox version 59.0)

Bookmarks Toolbar
  |
  |
  Development
    |
    |
    AWS
      |-> https://aws.amazon.com/
    Books
      |-> https://doc.rust-lang.org/book/
   Fiction
    |
    |
    Libraries
      |-> https://www.gutenberg.org/
    Books
      |-> https://www.gutenberg.org/ebooks/910

Import into Brave

Resulted import is:
Bookmarks Toolbar
  |
  |
  Development
    |
    |
    AWS
      |-> https://aws.amazon.com/
    Books
      |-> https://doc.rust-lang.org/book/
      |-> https://www.gutenberg.org/ebooks/910
   Fiction
    |
    |
    Libraries
      |-> https://www.gutenberg.org/

Fiction/Books folder is missing and its content is imported to Development/Books instead.

**Pull request merged**